### ISSUE TITLE
linuxbrew on a cluster (native blas/lapack/mpi)

### ISSUE THREAD

**Issue by davydden** [Homebrew/linuxbrew#504]
i wonder if one can efficiently use linuxbrew on a cluster? 

i guess this boils down to several questions:

1. can native blas/lapack libraries be used (e.g. mkl)? it's not difficult to modify some formulae to introduce somethign like `--with-native-blas` and try to link against, say `lblas` and `llapack` with a hope that they are in the path, but what if their names are different? To my understanding that's exactly the case of MKL libraries. So i wonder if one should have something like `export HOMEBREW_BLAS=...` and `export HOMEBREW_LAPACK=...` in order to use native libraries throughout?

2. it seems that it's not a problem with mpi, i.e. native mpi installation will be picked up by linuxbrew during installation if there is an `:mpi` dependecy. I have not checked it, though. 

3. some control over compilation flags (`CFLAGS` and `FFLAGS`) may also be necessary to fine-tune compilation and make each installed library/package run optimally on a given machine. Do we have this control in linuxbrew?

@Rombur you may want to jump into the discussion to share your experience in installation of scientific software on a cluster and a wish-list for the package manager.

**Comment by Rombur** [comment]
1. The most annoying part of installing deal.II on a new cluster is to link blas and lapack with Trilinos. Usually even after you load the module, cmake cannot find blas and lapack so you need to give the path explicitly. When you want to use MKL, there is the added problem that you need to link to three libraries instead of two and that none of them is called blas or lapack.

2. In candi, the user has to export CC, CXX, FC, and FF to the mpi compiler. We could require something similar with linuxbrew.

3. This would be nice but maybe not necessary. If you want to control each aspect of the compilation you can do it by hand. What could be nice is if for example if I want to compile petsc with some specific flags, I can compile it myself and then tell linuxbrew to use that particular installation of petsc. This would be nice but I can live without it.

**Comment by sjackman** [comment]
I'm not familiar with compiling numeric libraries myself. That being said, the best way to accomplish this in my opinion is to install all the dependencies using Homebrew, and then have `--with-<dependency>` options for the formulas that can use them optionally. i.e. `brew install dealii --with-trillinos`

**Comment by sjackman** [comment]
These issues are pretty much the same on either Mac or Linux, so this discussion would be better suited to Homebrew/homebrew-science. Shall I move this issue there?

**Comment by davydden** [comment]
@sjackman you are right that the context of the discussion is about installing `dealii` (open source finite element library) together with its linear algebra dependencies (`petsc`, `slepc`, `trilinos`) on a cluster. 

But in general, what i was trying to ask, i think, is a broader question of how can one use `linuxbrew` on a cluster and make it pick up `lapack`, `blas` and `mpi`, already available there? I think it is more about mechanisms which are needed to be able to use efficiently `linuxbrew` as a package manager on a cluster.
As an example, non-efficient usage would be to compile `openmpi` and `opneblas` yourself with linuxbrew.  

As @Rombur mentioned above for clusters with Intel Math Kernel LIbrary (MKL) implementation of Lapack it becomes cumbersome as one needs to link against these libraries which are not even called `liblapack` and/or `libblas`. If that is to be done using Homebrew/Linuxbrew/Homebrew-science, then one needs a way to provided those paths. So that's what I was trying to point out in (1) above.

I think this issue is Linuxrew only, as on os-x there is no question on how to install scientific software in general and `dealii` in particular. It all works with Homebrew/Homebrew-science. So i think the discussion belongs here, to Linuxbrew forum.

**Comment by dctrud** [comment]
Some kind of method for linking against MKL would definitely be really nice. We have a Linux cluster and workstation environment where we build optimized environment modules of software, but linuxbrew is a great option for users to try software, try out newer versions etc. R is potentially the thing we'd have users wanting most if this was possible.

**Comment by davydden** [comment]
@sjackman to re-iterate on my naive solution: how about Linuxbrew introduces another variable `export HOMEBREW_LAPACK=-L/path/to/mkl -lmkl1 -lmkl2 -lmkl3`. If there is a formula which depends on lapack/blas, we modify it to read this variable and link against those libraries instead of `-framework Accelerate` which are normally hard-coded. One still have an option of `openblas`, in such a case the environment variable is ignored.  `brew doctor` checks that this variable is set and enourage the user to adjust it according to his environment. 
@dctrud @Rombur do you think this could work in theory?

p.s. if this solution is in place, i am fine to take care of some formulae in homebrew-science such as petsc, slepc, trilinos etc. should be doable.

**Comment by davydden** [comment]
as a proof of the concept, if I do in the command line (close to real-life example)
```
$ export HOMEBREW_BLASLAPACK="-L/soft/com/packages/intel/13/079/mkl -lmkl_intel_ilp64 -lmkl_core -lmkl_intel_thread -lpthread -lm"
```
I can then use it in the formula (just printout for simplicity):
```
ohai ENV["HOMEBREW_BLASLAPACK"]
```
which shows
```
==> -L/soft/com/packages/intel/13/079/mkl -lmkl_intel_ilp64 -lmkl_core -lmkl_intel_thread -lpthread -lm
```
This is something which one could use, for example, in https://github.com/Homebrew/homebrew-science/blob/master/superlu_dist.rb#L31 to define blas libraries instead of default `-framework Accelerate`
@sjackman do you see my point?

**Comment by sjackman** [comment]
Yes. Is Intel MKL commercial software, and could it be packaged for Homebrew?

**Comment by davydden** [comment]
@sjackman yes, MKL is commercial, but MKL is not the only implementation of blas/lapack for clusters. The bigger picture, i would say, is that there are 2 pieces of software on clusters which are not meant to be self-compiled, but rather used the versions provided by the system  - `:mpi`, and `blas/lapack`. 
 `linuxbrew` has a lot of potential to be used on clusters to install open-source scientific software. To my understanding, however, in order to do so there should be mechanisms to use system provided `mpi` and `blas/lapack`. The approach above should potentially work to let system administrators configure the linuxbrew installation for particular cluster to use the system-provided blas/lapack.

p.s. @dctrud @Rombur please correct me if I am misleading in any sense or if you want to further elaborate on why one should use `blas/lapack` and `mpi` provided by the system.

**Comment by sjackman** [comment]
It could be possible to set a up `:blas` or `:lapack` dependency similar to the `:mpi` dependency. See https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/requirements/mpi_requirement.rb

**Comment by sjackman** [comment]
What's the relationship between BLAS and LAPACK?

**Comment by Rombur** [comment]
BLAS implements some simple operations that are extensively used in LAPACK

**Comment by sjackman** [comment]
`homebrew/science/openblas` provides BLAS, and `homebrew/dupes/lapack` provides LAPACK?
Does Intel MKL provide a source-compatible interface that implements both of these libraries?

**Comment by davydden** [comment]
yes, MKL is just a computer vendor **optimized** implementation of Lapack and blas, see http://www.netlib.org/lapack/faq.html#_what_and_where_are_the_lapack_vendors_implementations
As you can see, there are a number of vendor optimized implementations (AMD, Cray, Sun, HP, Intel, etc).

here is a nice summary article about BLAS, LAPACK and ScaLAPACK http://glennklockwood.blogspot.de/2014/02/quantum-espresso-compiling-and-choice.html

>BLAS is the lowest-level library and provides subroutines that do basic vector operations.  Netlib provides a reference implementation of BLAS written in Fortran, but the big idea behind BLAS is to allow hardware vendors to provide highly tuned versions of the BLAS subroutines that obviate the need for application developers to worry about optimizing their linear algebra for every possible computer architecture on which the application might run.  This motivation is also what gave rise to the MPI standard, but unlike MPI, BLAS is not an actual standard.
>
>LAPACK builds upon BLAS and provides higher-level matrix operations such as diagonalization (i.e., solving for eigenvectors and eigenvalues) and inversion.  BLAS and LAPACK seem to be bundled together when actually implemented (e.g., IBM ESSL and Intel MKL both provide both optimized BLAS and LAPACK), but they provide two distinct layers of abstracting the mathematical complexity away from application developers.
>
>ScaLAPACK builds upon LAPACK and provides a set of subroutines (prefixed with the letter P) that are analogous to the subroutines provided by LAPACK.  The big difference is that ScaLAPACK uses MPI to parallelize these LAPACK routines, whereas LAPACK itself (and the underlying BLAS) are completely serial (e.g., Netlib's reference distribution) or rely on shared memory for parallelization (e.g., multithreaded).

to sum it up, for HPC one actually use vendor optimized: `MPI`, `BLAS`, `LAPACK`, `ScaLAPACK`, `FFTs`. I can't comment on the last two, as they are not important in my work. But MPI, BLAS and LAPACK surely are. That's why i wanted to initiate the discussion on that topic.

**Comment by davydden** [comment]
btw, `homebrew/science/openblas` provides both `BLAS` and `LAPACK` (so no need for `homebrew/dupes/lapack`).  That's why it is currently used in numerous scientific formula such as `petsc`, `slepc`, `trilinos`, `superlu`, etc.

**Comment by davydden** [comment]
i think something like `:blas` and `:lapack` would make a lot of sense, not sure if those should be separate as often they are bundled together (either in vendor optimised versions or open-source projects such as `openblas` or `gotoblas2`[1] or Automatically Tuned Linear Algebra Software `atlas` [2] ).

[1] include most commonly used LAPACK functions
[2] a few routines from LAPACK

**Comment by sjackman** [comment]
@dpo Hi, Dominique. Do you have any comments on this thread? I don't use numerical algorithms much in my work, so I'm out of my zone here.

**Comment by davydden** [comment]
@tamiko you may want to comment on packaging as deal.ii developers currently face the issue of choosing the packaging system.

Perhaps Homebrew developers (@mikemcquaid @tdsmith ) could also join the discussion as it would be nice to have `:blas` and `:lapack` in the upstream even though for OSX one can hardcode `-framework Accelerate` where needed.

**Comment by tdsmith** [comment]
I would be curious to hear some stories about the different kinds of users Linuxbrew is intended to support and how they relate to the packages the base system is expected to provide.

I think that optionally providing alternative BLAS and LAPACK implementations is not interesting on OS X; openblas is not generally faster than Accelerate, and MKL is expensive and narrowly deployed on OS X clients.

I think a proper solution here could involve removing all optional openblas dependencies, forcing a build against Accelerate on OS X.

I'm warm to the idea of a special Linuxbrew-only :blas dependency (used like `depends_on :blas if OS.linux?`) with reasonable defaults (ideally: system BLAS if it's available, with fallback to a brewed openblas) and which may be overridden by an environment variable. But I probably won't implement it. :)

**Comment by sjackman** [comment]
I won't implement it either, but I'll review a pull request if someone wants to cook one up. As I said, I'm not a big user of numerical stuffs, so you'd be best off to find someone in `Homebrew/science` who is and can review this proposal properly. Perhaps @dpo

**Comment by davydden** [comment]
any1 on the homebrew team who do high performance computing ?
HPC => more developers choose Homebrew => more maintainers and better quality formulas => more users.

**Comment by dpo** [comment]
I am very favorable to adding BLAS/LAPACK support in the form of a `:blas` dependency. The latter could default to `-lblas -llapack`. Linux users who install BLAS/LAPACK with `apt-get` or similar will usually end up with `libblas.so` and `liblapack.so` in `/usr/lib` or another system path. On OSX, linking against `libblas.dylib` and `liblapack.dylib` is the same as linking against the `Accelerate` framework:
```
lrwxr-xr-x  1 root  wheel   116B Sep 20  2014 /usr/lib/libblas.dylib -> ../../System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
lrwxr-xr-x  1 root  wheel   118B Sep 20  2014 /usr/lib/liblapack.dylib -> ../../System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib
```

OpenBLAS lags behind `Accelerate` on OSX, at least on certain types of architecture. I opened [an issue](https://github.com/xianyi/OpenBLAS/issues/533) about that on the OpenBLAS tracker a while ago.

I'm unsure Homebrew core would accept several such dependencies, such as for example, `:blas`, `:mkl`, `:acml`, `:openblas`, `:gotoblas`, etc., though it might be best in the long run.

A simpler solution in the immediate future might be an option of the form
```ruby
option "with-blas=", "Specify desired BLAS library to link against"
```
and in `install`:
```ruby
blas = ARGV.value("with-blas") || "-lblas -llapack"  # or something of this form
```
We're already using this stanza in formulas that relate to MATLAB, e.g., [`suite-sparse`](https://github.com/Homebrew/homebrew-science/blob/master/suite-sparse.rb#L15), because a user might have several versions of Matlab available (I do).

The main difficulty is to propagate the BLAS choice to dependencies.

I'm willing to have a shot at a `:blas` dependency but it will likely not happen immediately.

**Comment by dctrud** [comment]
@dpo a `:blas` would be great, but getting the `with-blas` into formulae would be good. Given the myriad of schemes of environment modules and non-standard paths to access stuff on HPC systems I'd hazard it's probably not worth the effort to have specific mkl, atlas etc. deps/options. The relevant `-I` `-l` strings are likely to differ wildly for people most likely to need this on those kinds of systems. It's reasonable to expect enough local knowledge available to users to obtain the correct info?!

I'll try and find a little time to do something for a --with-blas on the R formula if I can. On our systems there is a massive difference between R configured as standard, and built with MKL (also intel compiler with full AVX opts in our case so wouldn't expect quite the same with gcc). For some simple linear algebra tests on a cluster node:

```
Task                               Standard R     Optimized R      Speedup
==========================================================================
Matrix Multiplication                  139.15         1.80          77x
Cholesky Decomposition                  29.53         0.32          61x
SVD                                     45.66         1.95          23x
PCA                                    201.30         6.25          32x
LDA                                    135.37        17.60           7x
```

**Comment by tdsmith** [comment]
I wouldn't generally like to see a proliferation of --with-blas options, especially because of the dependency propagation issues. I don't think I understand how formulas would use additional BLAS dependency symbols.

It's probably not tuned as nicely as MKL but fwiw openblas seems to be much better than Rblas.

Here is a sample C program we can compile to test for the presence of a system BLAS:
```c
double cblas_ddot(const int, const double*, const int, const double*, const int);
int main() { double x[] = {1.0, 2.0, 3.0}, y[] = {4.0, 5.0, 6.0}; cblas_ddot(3, x, 1, y, 1); return 0; }
```

**Comment by davydden** [comment]
@dpo i am glad to hear that your are interested in adding blas/lapack support. I agree that it would be good to have `:blas` in the upstream as well. Especially given the fact that on OSX `libblas` and `libblapack` symlink to accelerated framework, so this part can be unified between OS-X and Linux for all formulae in `homebrew-science` and `homebrew` itself. As far as blas/lapack is concerned, everything that worked for os-x will also work on linuxbrew, no need for extra if's of modification of formulae (linuxbrew vs homebrew).

I suppose if there is `:blas` (even for linuxbrew only) with optional HOMEBREW environment variables to override the default linker string (`-L/usr/lib/ -lblas -llapack`) and perhaps optional compiler line (e.g. `-I${MKLROOT}/include`), then it won't be a problem to propagate the blas choice, as effectively all formulas will be compiled against the same blas, as long as it is set up after installation. And even if someone would like to have different stacks of software compiled with different blas available on the system, this can be achieved using different `linuxbrew` installations sitting side by side on the system.

**Comment by tdsmith** [comment]
> But I probably won't implement it. :)

I lied; check out https://github.com/Homebrew/linuxbrew/pull/511.

I don't know what would need to happen to use pgcc or icc with Linuxbrew.

**Comment by sjackman** [comment]
Thanks, Tim. I'd love for `blas_requirement.rb` to live in Homebrew rather than Linuxbrew. Is there a use case for it on Mac, for example, a reasonably popular alternative to Accelerate framework?

**Comment by dpo** [comment]
The MKL, Atlas, and others are also available on the Mac.

**Comment by sjackman** [comment]
Great. Would you like to separate out `blas_requirement.rb` and send that upstream to Homebrew?

**Comment by davydden** [comment]
in case someone will come across this discussion, here is a small repo for experimenting with `blas` and `cmake` requirements https://github.com/davydden/homebrew-dealiisuite .
